arch/arm/src/stm32wl5/stm32wl5_ipcc.c

```diff
@@ -0,0 +1,801 @@
+      .txlen = IPCC_CHAN5_TX_SIZE - sizeof(unsigned)
+    }
+#endif
+
+#if IPCC_CHAN6
+    ,
+    {
+      .rxmem = (char *)(IPCC_CHAN6_START),
+      .rxlen = IPCC_CHAN6_RX_SIZE - sizeof(unsigned),
+      .txmem = (char *)(IPCC_CHAN6_START + IPCC_CHAN6_RX_SIZE),
+      .txlen = IPCC_CHAN6_TX_SIZE - sizeof(unsigned)
+    }
+#endif
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: stm32wl5_ipcc_tx_isr
+ *
+ * Description:
+ *   IPCC TX interrupt service routine. This interrupt is called when
+ *   second CPU read all data from IPCC TX memory and we are free to
+ *   write new data to that memory.
+ *
+ *   For buffered IPCC, we will immediately write data from buffer to
+ *   IPCC memory and notify second CPU. Also blocked writer will be
+ *   notified that there is free space on tx buffer.
+ *
+ *   For unbuffered IPCC, we won't copy anything but only notify blocked
+ *   writers that IPCC TX memory is free.
+ *
+ * Input Parameters:
+ *   irq - Number of IRQ that generated interrupt
+ *   context - Interrupt register state save info (architecture-specific)
+ *   arg - user data (not used by us)
+ *
+ * Returned Value:
+ *   Always OK
+ *
+ * Assumptions/Limitations:
+ *
+ ****************************************************************************/
+
+static int stm32wl5_ipcc_tx_isr(int irq, void *context, void *arg)
+{
+  int chan;
+  size_t nwritten;
+  uint32_t mr;
+  uint32_t sr;
+  uint32_t status;
+  struct stm32wl5_ipcc_s *priv;
+  struct stm32wl5_ipcc_chan_mem_s *txmem;
+
+  (void)context;
+  (void)arg;
+  (void)irq;
+
+  /* Disable TX interrupt, we don't want two "threads" to concurently
+   * modify circbuffers as we don't use any guard to them
+   */
+
+  up_disable_irq(STM32WL5_IRQ_IPCC_C1_TX_IT);
+
+  mr = getreg32(STM32WL5_IPCC_C1MR) >> STM32WL5_IPCC_TX_SHIFT;
+  sr = getreg32(STM32WL5_IPCC_C1TOC2SR);
+
+  /* Consider only channels that have tx memory free and are unmasked */
+
+  status = sr | mr;
+
+  /* Check which channels have data ready to read */
+
+  for (chan = 0; chan != IPCC_NCHAN; chan++)
+    {
+      if (status & (1 << chan))
+        {
+          /* Transmit on channel is either masked or TX memory is not
+           * ready to write (second CPU is still read data from it,
+           * CHnF == 1).
+           */
+
+          continue;
+        }
+
+      /* Get internal data structure for current chan */
+
+      priv = &g_ipccpriv[chan];
+      txmem = (struct stm32wl5_ipcc_chan_mem_s *)priv->txmem;
+
+#if CONFIG_IPCC_BUFFERED
+      /* Copy as much as we can into IPCC memory, circbuf won't copy
+       * more than there is in the buffer.
+       */
+
+      nwritten = circbuf_read(&priv->ipcc->txbuf, txmem->data, priv->txlen);
```

Comment:
Circbuf is very lightweight, it does couple int comparision then call memcpy(). Is copying couple of hundreds bytes that much for interrupt handler? I guess it might be if you copy a lot of data. I guess it would be best to install DMA handler here, which I might consider if I hit bottleneck here.

Reply:
I think the high priority work queue is more suitable for this, but consider this just as an option. Also I understand it needs more efforts to implement this, so let's keep it as is.